File: curriculum/units/M15/code/python/tool_scheduler.py
from __future__ import annotations

import asyncio
import inspect
from collections.abc import Awaitable, Callable, Mapping, Sequence
from dataclasses import dataclass
from types import MappingProxyType
from typing import Literal


@dataclass(frozen=True)
class Call:
    call_id: str
    name: str
    tool_input: Mapping[str, object]


@dataclass(frozen=True)
class Tool:
    name: str
    parse: Callable[[Mapping[str, object]], Mapping[str, object]]
    is_concurrency_safe: Callable[[Mapping[str, object]], bool]
    permission: Callable[[Mapping[str, object]], Literal["allow", "deny"]]
    run: Callable[
        [Mapping[str, object], Callable[[str], None], asyncio.Event],
        tuple[str, Mapping[str, str] | None]
        | Awaitable[tuple[str, Mapping[str, str] | None]],
    ]


@dataclass(frozen=True)
class Batch:
    mode: Literal["concurrent", "exclusive"]
    calls: tuple[Call, ...]


@dataclass(frozen=True)
class Outcome:
    call_id: str
    status: Literal["success", "error", "denied", "cancelled"]
    output: str
    update: Mapping[str, str] | None = None


class Scheduler:
    def __init__(self, tools: Sequence[Tool], limit: int = 2) -> None:
        if isinstance(limit, bool) or not isinstance(limit, int) or limit < 1:
            raise ValueError("limit must be positive")
        self._tools = {tool.name: tool for tool in tools}
        self._limit = limit

    def plan(self, calls: Sequence[Call]) -> tuple[Batch, ...]:
        batches: list[Batch] = []
        safe: list[Call] = []

        def flush() -> None:
            if safe:
                batches.append(Batch("concurrent", tuple(safe)))
                safe.clear()

        for call in calls:
            if self._classify(call):
                safe.append(call)
            else:
                flush()
                batches.append(Batch("exclusive", (call,)))
        flush()
        return tuple(batches)

    async def execute(
        self,
        calls: Sequence[Call],
        cancelled: asyncio.Event,
        on_progress: Callable[[str, str], None] = lambda _call_id, _stage: None,
    ) -> tuple[tuple[Outcome, ...], Mapping[str, str]]:
        outcomes: dict[str, Outcome] = {}
        context: dict[str, str] = {}
        for batch in self.plan(calls):
            if batch.mode == "exclusive":
                completed = (
                    await self._one(batch.calls[0], cancelled, on_progress),
                )
            else:
                completed = await self._concurrent(
                    batch.calls, cancelled, on_progress
                )
            outcomes.update((outcome.call_id, outcome) for outcome in completed)
            for call in batch.calls:
                update = outcomes[call.call_id].update
                if update is not None:
                    context.update(update)
        return tuple(outcomes[call.call_id] for call in calls), MappingProxyType(context)

    def _classify(self, call: Call) -> bool:
        tool = self._tools.get(call.name)
        if tool is None:
            return False
        try:
            parsed = tool.parse(call.tool_input)
            return tool.is_concurrency_safe(parsed)
        except Exception:
            return False

    async def _concurrent(
        self,
        calls: tuple[Call, ...],
        cancelled: asyncio.Event,
        on_progress: Callable[[str, str], None],
    ) -> tuple[Outcome, ...]:
        outcomes: list[Outcome | None] = [None] * len(calls)
        next_index = 0
        lock = asyncio.Lock()

        async def worker() -> None:
            nonlocal next_index
            while True:
                async with lock:
                    index = next_index
                    next_index += 1
                if index >= len(calls):
                    return
                outcomes[index] = await self._one(
                    calls[index], cancelled, on_progress
                )

        await asyncio.gather(
            *(worker() for _ in range(min(self._limit, len(calls))))
        )
        return tuple(outcome for outcome in outcomes if outcome is not None)

    async def _one(
        self,
        call: Call,
        cancelled: asyncio.Event,
        on_progress: Callable[[str, str], None],
    ) -> Outcome:
        if cancelled.is_set():
            return Outcome(call.call_id, "cancelled", "cancelled before execution")
        tool = self._tools.get(call.name)
        if tool is None:
            return Outcome(call.call_id, "error", f"unknown tool: {call.name}")
        try:
            parsed = tool.parse(call.tool_input)
        except Exception as error:
            return Outcome(call.call_id, "error", str(error))
        if tool.permission(parsed) == "deny":
            return Outcome(call.call_id, "denied", "permission denied")
        try:
            produced = tool.run(
                parsed,
                lambda stage: on_progress(call.call_id, stage),
                cancelled,
            )
            output, update = (
                await produced if inspect.isawaitable(produced) else produced
            )
            if cancelled.is_set():
                return Outcome(call.call_id, "cancelled", "cancelled during execution")
            return Outcome(call.call_id, "success", output, update)
        except Exception as error:
            return Outcome(
                call.call_id,
                "cancelled" if cancelled.is_set() else "error",
                str(error),
            )


def compare_snapshot_policies(
    calls: Sequence[tuple[bool, Mapping[str, str] | None]],
) -> tuple[Mapping[str, str], Mapping[str, str]]:
    response_complete: dict[str, str] = {}
    streaming: dict[str, str] = {}
    for safe, update in calls:
        if update is None:
            continue
        response_complete.update(update)
        if not safe:
            streaming.update(update)
    return MappingProxyType(response_complete), MappingProxyType(streaming)

File: curriculum/units/M15/code/typescript/tool-scheduler.ts
export type Call = Readonly<{
  id: string
  name: string
  input: Readonly<Record<string, unknown>>
}>

export type Progress = Readonly<{ callId: string; stage: string }>
export type ContextUpdate = Readonly<Record<string, string>>

export type Tool = Readonly<{
  name: string
  parse(input: Readonly<Record<string, unknown>>): Readonly<Record<string, unknown>>
  isConcurrencySafe(input: Readonly<Record<string, unknown>>): boolean
  permission(input: Readonly<Record<string, unknown>>): 'allow' | 'deny'
  run(
    input: Readonly<Record<string, unknown>>,
    report: (stage: string) => void,
    signal: AbortSignal,
  ): Promise<Readonly<{ output: string; update?: ContextUpdate }>>
}>

export type Batch = Readonly<{
  mode: 'concurrent' | 'exclusive'
  calls: readonly Call[]
}>

export type Outcome = Readonly<{
  callId: string
  status: 'success' | 'error' | 'denied' | 'cancelled'
  output: string
  update?: ContextUpdate
}>

export class Scheduler {
  readonly #tools: ReadonlyMap<string, Tool>
  readonly #limit: number

  constructor(tools: readonly Tool[], limit = 2) {
    this.#tools = new Map(tools.map(tool => [tool.name, tool]))
    if (!Number.isInteger(limit) || limit < 1) throw new Error('limit must be positive')
    this.#limit = limit
  }

  plan(calls: readonly Call[]): readonly Batch[] {
    const batches: Batch[] = []
    let safe: Call[] = []
    const flush = (): void => {
      if (safe.length === 0) return
      batches.push(Object.freeze({ mode: 'concurrent', calls: Object.freeze(safe) }))
      safe = []
    }
    for (const call of calls) {
      if (this.#classify(call)) {
        safe.push(call)
      } else {
        flush()
        batches.push(Object.freeze({ mode: 'exclusive', calls: Object.freeze([call]) }))
      }
    }
    flush()
    return Object.freeze(batches)
  }

  async execute(
    calls: readonly Call[],
    signal: AbortSignal,
    onProgress: (progress: Progress) => void = () => {},
  ): Promise<Readonly<{ outcomes: readonly Outcome[]; context: Readonly<Record<string, string>> }>> {
    const outcomes = new Map<string, Outcome>()
    let context: Readonly<Record<string, string>> = Object.freeze({})
    for (const batch of this.plan(calls)) {
      const completed = batch.mode === 'exclusive'
        ? [await this.#one(batch.calls[0]!, signal, onProgress)]
        : await this.#concurrent(batch.calls, signal, onProgress)
      for (const outcome of completed) outcomes.set(outcome.callId, outcome)
      for (const call of batch.calls) {
        const update = outcomes.get(call.id)?.update
        if (update) context = Object.freeze({ ...context, ...update })
      }
    }
    return Object.freeze({
      outcomes: Object.freeze(calls.map(call => outcomes.get(call.id)!)),
      context,
    })
  }

  #classify(call: Call): boolean {
    const tool = this.#tools.get(call.name)
    if (!tool) return false
    try {
      const parsed = tool.parse(call.input)
      return tool.isConcurrencySafe(parsed)
    } catch {
      return false
    }
  }

  async #concurrent(
    calls: readonly Call[],
    signal: AbortSignal,
    onProgress: (progress: Progress) => void,
  ): Promise<readonly Outcome[]> {
    const outcomes = new Array<Outcome>(calls.length)
    let next = 0
    const worker = async (): Promise<void> => {
      while (true) {
        const index = next++
        if (index >= calls.length) return
        outcomes[index] = await this.#one(calls[index]!, signal, onProgress)
      }
    }
    await Promise.all(Array.from({ length: Math.min(this.#limit, calls.length) }, worker))
    return Object.freeze(outcomes)
  }

  async #one(
    call: Call,
    signal: AbortSignal,
    onProgress: (progress: Progress) => void,
  ): Promise<Outcome> {
    if (signal.aborted) return cancelled(call.id, 'cancelled before execution')
    const tool = this.#tools.get(call.name)
    if (!tool) return error(call.id, `unknown tool: ${call.name}`)
    let parsed: Readonly<Record<string, unknown>>
    try {
      parsed = tool.parse(call.input)
    } catch (reason) {
      return error(call.id, message(reason))
    }
    if (tool.permission(parsed) === 'deny') {
      return Object.freeze({ callId: call.id, status: 'denied', output: 'permission denied' })
    }
    try {
      const result = await tool.run(
        parsed,
        stage => onProgress(Object.freeze({ callId: call.id, stage })),
        signal,
      )
      if (signal.aborted) return cancelled(call.id, message(signal.reason))
      return Object.freeze({
        callId: call.id,
        status: 'success',
        output: result.output,
        ...(result.update ? { update: result.update } : {}),
      })
    } catch (reason) {
      return signal.aborted
        ? cancelled(call.id, message(reason))
        : error(call.id, message(reason))
    }
  }
}

export function compareSnapshotPolicies(
  calls: readonly Readonly<{ id: string; safe: boolean; update?: ContextUpdate }>[]
): Readonly<{
  responseCompleteContext: Readonly<Record<string, string>>
  streamingContext: Readonly<Record<string, string>>
}> {
  let responseCompleteContext: Readonly<Record<string, string>> = Object.freeze({})
  let streamingContext: Readonly<Record<string, string>> = Object.freeze({})
  for (const call of calls) {
    if (!call.update) continue
    responseCompleteContext = Object.freeze({ ...responseCompleteContext, ...call.update })
    if (!call.safe) streamingContext = Object.freeze({ ...streamingContext, ...call.update })
  }
  return Object.freeze({ responseCompleteContext, streamingContext })
}

function cancelled(callId: string, output: string): Outcome {
  return Object.freeze({ callId, status: 'cancelled', output })
}

function error(callId: string, output: string): Outcome {
  return Object.freeze({ callId, status: 'error', output })
}

function message(reason: unknown): string {
  return reason instanceof Error ? reason.message : String(reason ?? 'cancelled')
}
